File: backend/functions/src/task_events.ts
import * as admin from "firebase-admin";

export interface ExternalTaskEvent {
  userId: string;
  providerId: string;
  entityType: string;
  externalId: string;
  date: string; // YYYY-MM-DD format (CivilDay)
  action: "completed" | "uncompleted" | "dismissed";
  timestamp?: string;
  metadata?: Record<string, unknown>;
}

export interface TaskEventResult {
  success: boolean;
  instanceId?: string;
  actionApplied: string;
  createdNewInstance?: boolean;
  message?: string;
}

/**
 * Validates the structure of an incoming ExternalTaskEvent.
 */
export function validateTaskEvent(payload: unknown): { valid: boolean; error?: string; event?: ExternalTaskEvent } {
  if (!payload || typeof payload !== "object") {
    return { valid: false, error: "Event payload must be a non-null object" };
  }

  const p = payload as Record<string, unknown>;
  const requiredFields = ["userId", "providerId", "entityType", "externalId", "date", "action"];
  for (const field of requiredFields) {
    if (typeof p[field] !== "string" || !p[field]) {
      return { valid: false, error: `Missing or invalid required string field: ${field}` };
    }
  }

  const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
  if (!dateRegex.test(p.date as string)) {
    return { valid: false, error: "Field 'date' must match YYYY-MM-DD format" };
  }

  const validActions = ["completed", "uncompleted", "dismissed"];
  if (!validActions.includes(p.action as string)) {
    return { valid: false, error: `Invalid action '${p.action}'. Must be one of: ${validActions.join(", ")}` };
  }

  return {
    valid: true,
    event: {
      userId: p.userId as string,
      providerId: p.providerId as string,
      entityType: p.entityType as string,
      externalId: p.externalId as string,
      date: p.date as string,
      action: p.action as "completed" | "uncompleted" | "dismissed",
      timestamp: typeof p.timestamp === "string" ? p.timestamp : new Date().toISOString(),
      metadata: typeof p.metadata === "object" && p.metadata !== null ? (p.metadata as Record<string, unknown>) : undefined,
    },
  };
}

/**
 * Processes an incoming ExternalTaskEvent and updates/creates the corresponding TaskInstance in Firestore.
 */
export async function processExternalTaskEvent(
  db: admin.firestore.Firestore,
  event: ExternalTaskEvent
): Promise<TaskEventResult> {
  const instancesRef = db.collection("users").doc(event.userId).collection("instances");

  // Query instances matching the date and provider/external ID binding
  const querySnap = await instancesRef
    .where("scheduledDate", "==", event.date)
    .where("integrationBinding.providerId", "==", event.providerId)
    .where("integrationBinding.externalId", "==", event.externalId)
    .limit(1)
    .get();

  const now = new Date();
  const nowIso = now.toISOString();

  let targetStatus = "pending";
  let completedAt: string | null = null;
  let completedByUserId: string | null = null;

  if (event.action === "completed") {
    targetStatus = "completed";
    completedAt = event.timestamp || nowIso;
    completedByUserId = event.userId;
  } else if (event.action === "dismissed") {
    targetStatus = "dismissed";
  } else if (event.action === "uncompleted") {
    targetStatus = "pending";
    completedAt = null;
    completedByUserId = null;
  }

  if (!querySnap.empty) {
    // Update existing instance
    const doc = querySnap.docs[0];
    await doc.ref.update({
      status: targetStatus,
      completedAt: completedAt,
      completedByUserId: completedByUserId,
      updatedAt: nowIso,
      lastModifiedByUserId: event.userId,
    });

    return {
      success: true,
      instanceId: doc.id,
      actionApplied: event.action,
      createdNewInstance: false,
    };
  }

  // If no instance exists for that day, check if a parent TaskSchedule template exists
  const schedulesRef = db.collection("users").doc(event.userId).collection("tasks");
  const scheduleSnap = await schedulesRef
    .where("integrationBinding.providerId", "==", event.providerId)
    .where("integrationBinding.externalId", "==", event.externalId)
    .limit(1)
    .get();

  let scheduleId = `SCHED-${event.providerId}-${event.externalId}`;
  let title = `${event.providerId}: ${event.externalId}`;
  let description = `Auto-tracked from ${event.providerId}`;

  if (!scheduleSnap.empty) {
    const sDoc = scheduleSnap.docs[0];
    scheduleId = sDoc.id;
    const sData = sDoc.data();
    if (sData.title) title = sData.title;
    if (sData.description) description = sData.description;
  }

  // Create a Just-in-Time TaskInstance document
  const newInstanceRef = instancesRef.doc();
  const newInstance = {
    id: newInstanceRef.id,
    scheduleId: scheduleId,
    ruleId: "RULE-EXT-SYNC",
    title: title,
    description: description,
    scheduledDate: event.date,
    startRelativeTime: { minutes: 0 },
    dueRelativeTime: { minutes: 1439 },
    isFamily: false,
    status: targetStatus,
    completedAt: completedAt,
    completedByUserId: completedByUserId,
    completedByUserIds: completedByUserId ? [completedByUserId] : [],
    integrationBinding: {
      providerId: event.providerId,
      entityType: event.entityType,
      externalId: event.externalId,
      bidirectional: true,
    },
    updatedAt: nowIso,
    createdAt: nowIso,
    lastModifiedByUserId: event.userId,
  };

  await newInstanceRef.set(newInstance);

  return {
    success: true,
    instanceId: newInstanceRef.id,
    actionApplied: event.action,
    createdNewInstance: true,
    message: `Created and applied ${event.action} to new TaskInstance`,
  };
}

File: backend/functions/lib/src/handlers/task_events.dart
import '../interop/firebase_admin.dart';
import '../interop/firebase_functions.dart';
import '../interop/node_interop.dart';
import '../models/civil_day.dart';
import '../models/task_event.dart';

/// Validates the structure of an incoming ExternalTaskEvent payload.
TaskEventValidation validateTaskEvent(dynamic payload) {
  if (payload == null || payload is! Map) {
    return const TaskEventValidation(
      valid: false,
      error: 'Event payload must be a non-null object',
    );
  }

  final p = Map<String, dynamic>.from(payload);
  const requiredFields = [
    'userId',
    'providerId',
    'entityType',
    'externalId',
    'date',
    'action',
  ];

  for (final field in requiredFields) {
    final val = p[field];
    if (val is! String || val.isEmpty) {
      return TaskEventValidation(
        valid: false,
        error: 'Missing or invalid required string field: $field',
      );
    }
  }

  final dateStr = p['date'] as String;
  if (!CivilDay.isValid(dateStr)) {
    return const TaskEventValidation(
      valid: false,
      error: "Field 'date' must match YYYY-MM-DD format",
    );
  }

  const validActions = ['completed', 'uncompleted', 'dismissed'];
  final action = p['action'] as String;
  if (!validActions.contains(action)) {
    return TaskEventValidation(
      valid: false,
      error:
          "Invalid action '$action'. Must be one of: ${validActions.join(', ')}",
    );
  }

  return TaskEventValidation(
    valid: true,
    event: ExternalTaskEvent(
      userId: p['userId'] as String,
      providerId: p['providerId'] as String,
      entityType: p['entityType'] as String,
      externalId: p['externalId'] as String,
      date: dateStr,
      action: action,
      timestamp: p['timestamp'] is String
          ? p['timestamp'] as String
          : DateTime.now().toUtc().toIso8601String(),
      metadata: p['metadata'] is Map
          ? Map<String, dynamic>.from(p['metadata'] as Map)
          : null,
    ),
  );
}

/// Authenticates an incoming request for reporting task events.
/// Accepts either:
/// 1. A valid Firebase Auth ID token matching the event userId (or with admin claims).
/// 2. A valid service secret matching TASK_HUB_SECRET or SERVICE_SECRET environment variable.
Future<TaskAuthValidationResult> authenticateTaskEventRequest(
  Map<String, dynamic> headers,
  String targetUserId, {
  AuthService? auth,
  String? envSecret,
}) async {
  final authHeaderRaw = headers['authorization'] ?? headers['Authorization'];
  final authHeader = authHeaderRaw is List
      ? (authHeaderRaw.isNotEmpty ? authHeaderRaw.first.toString() : null)
      : authHeaderRaw?.toString();

  final serviceKeyRaw = headers['x-service-secret'] ?? headers['x-api-key'];
  final serviceKey = serviceKeyRaw is List
      ? (serviceKeyRaw.isNotEmpty ? serviceKeyRaw.first.toString() : null)
      : serviceKeyRaw?.toString();

  final secret =
      envSecret ?? getEnv('TASK_HUB_SECRET') ?? getEnv('SERVICE_SECRET');

  // 1. Check Service Secret Header or Bearer secret
  if (secret != null &&
      secret.isNotEmpty &&
      (serviceKey == secret || authHeader == 'Bearer $secret')) {
    return const TaskAuthValidationResult(authenticated: true);
  }

  // 2. Check Firebase ID Token in Bearer header
  if (authHeader != null && authHeader.startsWith('Bearer ')) {
    final idToken = authHeader.substring(7).trim();
    if (idToken.isNotEmpty) {
      try {
        final firebaseAuth = auth ?? getFirebaseAdminAuth();
        final decodedToken = await firebaseAuth.verifyIdToken(idToken);
        if (decodedToken.uid == targetUserId || decodedToken.admin == true) {
          return const TaskAuthValidationResult(authenticated: true);
        }
        return const TaskAuthValidationResult(
          authenticated: false,
          status: 403,
          error: 'Forbidden: Authenticated user does not match target userId.',
        );
      } catch (_) {
        return const TaskAuthValidationResult(
          authenticated: false,
          status: 401,
          error: 'Unauthorized: Invalid or expired authentication token.',
        );
      }
    }
  }

  return const TaskAuthValidationResult(
    authenticated: false,
    status: 401,
    error: 'Unauthorized: Missing or invalid authentication credentials.',
  );
}

/// Processes an incoming ExternalTaskEvent and updates/creates the corresponding TaskInstance in Firestore.
Future<TaskEventResult> processExternalTaskEvent(
  FirestoreDatabase db,
  ExternalTaskEvent event, {
  DateTime? now,
}) async {
  final instancesRef =
      db.collection('users').doc(event.userId).collection('instances');

  // Query instances matching the date and provider/external ID binding
  final querySnap = await instancesRef
      .where('scheduledDate', '==', event.date)
      .where('integrationBinding.providerId', '==', event.providerId)
      .where('integrationBinding.externalId', '==', event.externalId)
      .limit(1)
      .get();

  final nowUtc = now ?? DateTime.now().toUtc();
  final nowIso = nowUtc.toIso8601String();

  var targetStatus = 'pending';
  String? completedAt;
  String? completedByUserId;

  if (event.action == 'completed') {
    targetStatus = 'completed';
    completedAt = event.timestamp ?? nowIso;
    completedByUserId = event.userId;
  } else if (event.action == 'dismissed') {
    targetStatus = 'dismissed';
  } else if (event.action == 'uncompleted') {
    targetStatus = 'pending';
    completedAt = null;
    completedByUserId = null;
  }

  if (!querySnap.empty) {
    // Update existing instance
    final doc = querySnap.docs.first;
    final existingData = doc.data() ?? {};
    final updatedCompletedByUserIds = List<String>.from(
      (existingData['completedByUserIds'] as List?) ?? [],
    );

    if (event.action == 'completed') {
      if (!updatedCompletedByUserIds.contains(event.userId)) {
        updatedCompletedByUserIds.add(event.userId);
      }
    } else if (event.action == 'uncompleted') {
      updatedCompletedByUserIds.removeWhere((id) => id == event.userId);
    }

    await doc.ref.update({
      'status': targetStatus,
      'completedAt': completedAt,
      'completedByUserId': completedByUserId,
      'completedByUserIds': updatedCompletedByUserIds,
      'updatedAt': nowIso,
      'lastModifiedByUserId': event.userId,
    });

    return TaskEventResult(
      success: true,
      instanceId: doc.id,
      actionApplied: event.action,
      createdNewInstance: false,
    );
  }

  // If no instance exists for that day, check if a parent TaskSchedule template exists
  final schedulesRef =
      db.collection('users').doc(event.userId).collection('tasks');
  final scheduleSnap = await schedulesRef
      .where('integrationBinding.providerId', '==', event.providerId)
      .where('integrationBinding.externalId', '==', event.externalId)
      .limit(1)
      .get();

  var scheduleId = 'SCHED-${event.providerId}-${event.externalId}';
  var title = '${event.providerId}: ${event.externalId}';
  var description = 'Auto-tracked from ${event.providerId}';

  if (!scheduleSnap.empty) {
    final sDoc = scheduleSnap.docs.first;
    scheduleId = sDoc.id;
    final sData = sDoc.data() ?? {};
    if (sData['title'] is String) title = sData['title'] as String;
    if (sData['description'] is String)
      description = sData['description'] as String;
  }

  // Create a Just-in-Time TaskInstance document
  final newInstanceRef = instancesRef.doc();
  final newInstance = <String, dynamic>{
    'id': newInstanceRef.id,
    'scheduleId': scheduleId,
    'ruleId': 'RULE-EXT-SYNC',
    'title': title,
    'description': description,
    'scheduledDate': event.date,
    'startRelativeTime': {'minutes': 0},
    'dueRelativeTime': {'minutes': 1439},
    'isFamily': false,
    'status': targetStatus,
    'completedAt': completedAt,
    'completedByUserId': completedByUserId,
    'completedByUserIds':
        completedByUserId != null ? [completedByUserId] : <String>[],
    'integrationBinding': {
      'providerId': event.providerId,
      'entityType': event.entityType,
      'externalId': event.externalId,
      'bidirectional': true,
    },
    'updatedAt': nowIso,
    'createdAt': nowIso,
    'lastModifiedByUserId': event.userId,
  };

  await newInstanceRef.set(newInstance);

  return TaskEventResult(
    success: true,
    instanceId: newInstanceRef.id,
    actionApplied: event.action,
    createdNewInstance: true,
    message: 'Created and applied ${event.action} to new TaskInstance',
  );
}

/// HTTP endpoint handler for reportExternalTaskEvent.
Future<void> handleReportExternalTaskEvent(
  HttpRequest req,
  HttpResponse res, {
  FirestoreDatabase? db,
  AuthService? auth,
}) async {
  if (req.method != 'POST') {
    res
        .status(405)
        .json({'success': false, 'error': 'Method Not Allowed. Use POST.'});
    return;
  }

  final validation = validateTaskEvent(req.body);
  if (!validation.valid || validation.event == null) {
    logWarn(
        'Invalid external task event received: ${req.body} ${validation.error}');
    res.status(400).json({'success': false, 'error': validation.error});
    return;
  }

  final authResult = await authenticateTaskEventRequest(
    req.headers,
    validation.event!.userId,
    auth: auth,
  );
  if (!authResult.authenticated) {
    logWarn(
      'Unauthorized external task event attempt for user ${validation.event!.userId}: ${authResult.error}',
    );
    res
        .status(authResult.status ?? 401)
        .json({'success': false, 'error': authResult.error});
    return;
  }

  try {
    final firestoreDb = db ?? getFirebaseAdminDb();
    final result =
        await processExternalTaskEvent(firestoreDb, validation.event!);
    logInfo(
      'Processed task event for user ${validation.event!.userId}, provider ${validation.event!.providerId}: ${validation.event!.action}',
    );
    res.status(200).json(result.toJson());
  } catch (error) {
    final errorMessage = error.toString().replaceFirst('Exception: ', '');
    logError('Error processing external task event:', error);
    res.status(500).json({'success': false, 'error': errorMessage});
  }
}
